crates/ruff_python_parser/src/lib.rs

```diff
@@ -610,6 +655,36 @@ impl Deref for Tokens {
     }
 }
 
+/// One challenge here is how to we keep the spine to the root.
+pub enum TokenAt {
+    /// There's no node at the given offset
+    None,
+
+    /// A single node fully encloses the offset.
+    Single(Token),
+
+    /// The offset falls exactly between two nodes. E.g. `CURSOR` in `call<CURSOR>(arguments)` is
+    /// positioned exactly between the callee and the arguments.
+    Between(Token, Token),
+}
+
+impl Iterator for TokenAt {
+    type Item = Token;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match std::mem::replace(self, TokenAt::None) {
+            TokenAt::None => None,
+            TokenAt::Single(token) => Some(token),
+            TokenAt::Between(first, second) => {
+                *self = TokenAt::Single(second);
+                Some(first)
+            }
+        }
```

Comment:
nit: maybe we can use the same logic as `TokenAt::Between` to avoid the `mem::replace` all for every iteration?

```suggestion
        match self {
            TokenAt::None => None,
            TokenAt::Single(token) => {
                *self = TokenAt::None;
                Some(token)
            }
            TokenAt::Between(first, second) => {
                *self = TokenAt::Single(second);
                Some(first)
            }
        }
```